### PR TITLE
Search for pyproject.toml config file in parent dirs

### LINKED ISSUE #3289: Looking for configuration files like ``pylintrc`` or ``pyproject.toml`` in parent directories like ``black`` or ``isort``
### Is your feature request related to a problem? Please describe
I use pylint in vi with a pyproject.toml -- pyproject.toml isn't always in my $CWD.

### Describe the solution you'd like

What do you think about also looking at ../../.. recursively like you do the .pylintrc?

### Additional context

You might want to exclude setup.cfg from this for backwards compatibility.

### PR DESCRIPTION
**Pull request by Felixoid**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description
Search for pyproject.toml in parent directories.

Closes #3289

### CODE REVIEW

**Comment by Felixoid** [comment]
How would I add exceptions for false-positive spell checker?

**Comment by DanielNoord** [comment]
Thanks for the interest in this issue @Felixoid. I didn't notice we left that issue open, but due to discussion in https://github.com/PyCQA/pylint/pull/3769 I am not sure if we can merge this PR.
Even if we do want to I think we should consider whether this is a breaking change and if it can be included in `2.x`.

**Comment by Pierre-Sassoulas** [comment]
I think this should use the specification from the discussion in https://github.com/PyCQA/pylint/issues/3289

**Comment by Pierre-Sassoulas** [comment]
> How would I add exceptions for false-positive spell checker?

There's a dictionary at the root of the project.

**Comment by Felixoid** [comment]
Hi @DanielNoord, there's nothing relatively new in the code that wasn't there already. The snippet `curdir = Path(os.getcwd()).resolve()` is a bit below, so it should work fine.

Dear @Pierre-Sassoulas, you mean that `.git/.hg` as another marker for the `pyproject.toml` directory? Or the max depth? Or rather that it should search for any possible configuration up to a repository root directory?

**Comment by Pierre-Sassoulas** [comment]
We handle multiple configuration file (``pylintrc``/``.pylintrc`` / ``pyproject.toml``, ``setup.cfg`` too) I think we should upgrade the current function to search for configuration files to make it search up the tree until it hits a configuration file or a ``.git`` / ``.hg`` directory. Unless I'm mistaken it's this function that need to be modified:

https://github.com/PyCQA/pylint/blob/2e2f8f8c12a2e54c18c0e4dd49e4850f5c913b1f/pylint/config/find_default_config_files.py#L42

**Comment by Felixoid** [comment]
Thank you for your review, I've added processing of all supported configs, and updated the docs

**Comment by Felixoid** [comment]
It definitely went out of control. I'll take care of it.

**Comment by DanielNoord** [comment]
The failing tests show that this is not a trivial change. Again, I'm against changing this in a `2.x` update. Consider:
 
A directory without a `git repo` and no `pyproject.toml`. For example, a collection of (internal) scripts.
An user can now run `pylint` on this directory and get the default values and pass the checks.
With this change we would now start going up until the user's root, thereby potentially changing the config file that is being used for their checks.

I'm usually fine with having users change some configuration stuff between minor patches, but forcing them to restructure their directories or move files around doesn't seem like a good thing to do in a minor version update.

**Comment by Felixoid** [comment]
Ok, I misinterpret your first message. I thought that you've meant python2.

You are right on one hand. On another, now nobody can set a project global pylintrc, of there is a mixed repo. That's why I started working on the issue, basically

**Comment by DanielNoord** [comment]
> Ok, I misinterpret your first message. I thought that you've meant python2.
> 
> You are right on one hand. On another, now nobody can set a project global pylintrc, of there is a mixed repo. That's why I started working on the issue, basically

See L62-80 of the original code. We currently support an environment variable and a global config in the users home directory. This probably doesn't cover all cases (and it might still make sense to merge this PR at some point), but I'm not sure we can change this now.

I haven't tested this myself but can't you also use `pylint --rcfile=../../../pylintrc`?

**Comment by DanielNoord** [comment]
> It's precisely the issue I am trying to solve. And to show that it works, we already have the best proof. Take a look on [the report](https://github.com/PyCQA/pylint/pull/7163#issuecomment-1180916413). Each tested project has less number of reports than it has. It works as expected.

This is an unrelated report and caused by a bug in our primer. Actually I would expect this PR to have no effect in the primer as all primer projects are linted with a specified `pylintrc`. For now we should ignore all primer comments in this PR until #7160 has been merged.

> What can be done, for now I'd roll back to the approximate state from [the commit](https://github.com/PyCQA/pylint/compare/d5f99f6560afff9d88857af3dcfa22f6807db691..486ab25477e04a88d85786cd81545d04afa9ee12#diff-a76d406b648f773c81008092021cb6142d2c4d8e3d3e5bf241e88944a2afb541L33) and add `.git`/`.hg` there only for `pyproject.toml` configuration. It's not such a significant change, and one is looking for a way to set a repo-wide config will have it.
> 
> The major change to search all configs in the directories tree may be then addressed in a major release.
> 
> Does it sound reasonable?

This is still a breaking change for those who want to lint subdirectories with default values. I can't stress enough the range of different `pylint` invocations. Take for example https://github.com/PyCQA/pylint/issues/6799, which shows that some users actually run `pylint` twice over the same repository with the same configuration file but with different command line arguments. Initially that seemed totally bizarre to me, but I guess people find a use case for it. Same goes for linting subdirectories with default values, we just don't know whether users rely on this.

I can see the need for repository wide configuration files, but I don't think this can be default value. Perhaps we can work with a flag? `--look-for-parent-configuration` or something?

**Comment by Felixoid** [comment]
Sorry, I can't call it breaking changes. For me, it is still an essential function that wasn't implemented for historical reasons.

The point is that each tool I can recall, e.g., black and isort (see the issue), [flake8](https://github.com/PyCQA/flake8/blob/ca573a7ccf2d4a1c7df0b577bb6d3455c941e828/src/flake8/options/config.py#L17-L31) already supports it at least somehow. 

I hear your point. I can only agree to disagree. I see people need the same thing as I do, so this PR solves the issue that has been hanging around for three years.

What I can do more here is write tests to put to stone the proposed behavior with `pyproject.toml or .git or .hg in a directory above`. I wouldn't invest my time to invest then even more time into wrapping around workarounds.

**Comment by DanielNoord** [comment]
This discussion has gotten stuck I think. I'm tagging some other active maintainers to gather more opinions about this.

@Pierre-Sassoulas @jacobtylerwalls

**Comment by jacobtylerwalls** [comment]
Hey @Felixoid, thanks for these changes. I think everyone is positive on your changes, we're just trying to figure out what release to put it in.

The good news is that even if it goes in 3.0, we'll probably put out a pre-release, and you can start using it. As I said [here](https://github.com/PyCQA/pylint/issues/3512#issuecomment-1173085153), merging great patches to a 3.0 branch just increases the urgency of releasing it. :-)

I see the argument on both sides. Honestly, this is what I thought the configurations were:

> I can see the need for repository wide configuration files

But @DanielNoord has a point that people's CI workflows will break given that often they don't even know what their `cwd` is. (Even in developing pylint itself recently we've had several issues that hinged on cwd-sensitivity.) And people can get cross when we make changes that affect how they invoke pylint.

I don't feel too strongly either way, but I just want to reiterate that this part of the discussion should slow down reviewing the changes and getting them merged either to `main` or to `3.0`.

**Comment by Pierre-Sassoulas** [comment]
> merging great patches to a 3.0 branch just increases the urgency of releasing it.

👍 Strong agree. We have a LOT to release in 3.0, we need to add things that sparkle joy or we'll never ship it 😄

But this is not the reason I'm slightly on the side of doing the feature as a breaking change here. Environment bugs are really nasty to tests, troubleshoot, release, and get right in general. We can't anticipate all the possible environments used (we had solaris and another really niche OS recently, there's probably a Mars Rover prototype with custom hardware somewhere too). This is probably simpler in this case, which is why I'm only slightly in favor, but there's a very high probability that we're going to break something somewhere and this is not good practice to take that risk in a minor (especially as pylint add new messages with new checkers and fixed false negatives, so the migration to a new minor is not painless already).

**Comment by Felixoid** [comment]
Thank you for your feedback, I got your point. I agree to redo it as a major change.

Unfortunately, I got serious sick today, and wouldn't be able to do it this week. Maybe, the weekend sounds realistic.

**Comment by Felixoid** [comment]
As far as I see, the PR currently has a base `main`, so I guess it will end up in the 3.0, right?

**Comment by Pierre-Sassoulas** [comment]
> As far as I see, the PR currently has a base main, so I guess it will end up in the 3.0, right?

We're putting everything from main in 3.0 right now, yes.

**Comment by Felixoid** [comment]
Hello again.

Is there anything preventing it from merging now?

**Comment by Pierre-Sassoulas** [comment]
I added the need review label, but I don't have the availability right now :)

**Comment by DanielNoord** [comment]
I think the main issue is that we don't really have branch to put this on. We recently removed the `3.x` branch as it took effort to keep it up to date while we are no where near releasing `3.x` currently.

**Comment by jacobtylerwalls** [comment]
That shouldn't stop a review, though. I think availability of a review is the immediate blocker.

**Comment by Pierre-Sassoulas** [comment]
Creating a 3.x branch from the current main branch should be easy once we need it (Ideally I would create this branch officially only when we really need it because then we'll have to merge main back in it for quite a long time)

**Comment by Felixoid** [comment]
I updated the branch, so it's adjusted to recent fixes. Can it process further, please?

File: pylint/config/find_default_config_files.py
Comment (Pierre-Sassoulas) [inline pylint/config/find_default_config_files.py:20]:
```suggestion
PYPROJECT_NAME = "pyproject.toml"
```

File: tests/config/test_find_default_config_files.py
Comment (Pierre-Sassoulas) [inline tests/config/test_find_default_config_files.py:149]: Shouldn't 'files' and 'expected' be in a ``@pytest.mark.parametrize`` instead ?

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Thank you for implementing this, I have a nitpick around the test implementation but LGTM.

File: tests/config/test_find_default_config_files.py
Comment (Felixoid) [reply tests/config/test_find_default_config_files.py:149]: I don't know, sorry. It's the same was as `test_pylintrc_parentdir` is tested RN. Can we consider refactoring both tests later, maybe?

Comment (Pierre-Sassoulas) [reply tests/config/test_find_default_config_files.py:149]: I did not realize that it's the exact same list than another test. We might want to refactor before the change and use the same constant in both, then ?

File: pylint/config/find_default_config_files.py
Comment (Pierre-Sassoulas) [reply pylint/config/find_default_config_files.py:20]: I missed that, it felt like we did not need a path in this particular case, but let's keep it that way.

File: tests/config/test_find_default_config_files.py
Comment (Felixoid) [reply tests/config/test_find_default_config_files.py:149]: It's not used anywhere else. It's just the same way of testing that is used in other places in this file. I don't see any other example of `files` and `expected` being a `@pytest.mark.parametrize`

See `test_pylintrc_parentdir`, `test_pylintrc_parentdir_no_package`, `test_verbose_output_no_config`, and `test_verbose_abbreviation`

Comment (Pierre-Sassoulas) [reply tests/config/test_find_default_config_files.py:149]: I'm going to resolve then.

**Review by Pierre-Sassoulas — APPROVED** [review]

**Comment by DanielNoord** [comment]
Thanks for the review @Pierre-Sassoulas.

As far as I can see though there is still no real decision on whether this should go in `2.16.0`. Imo, this is a hard to troubleshoot bug that can potentially break CIs. I really don't think the benefit it offers outweigh putting a breaking change in a minor patch.

**Comment by Pierre-Sassoulas** [comment]
Right I forgot about that and just added it automatically to the next release. We need to create the 3.0 branch and deal with the 3.0 deprecations then.

**Comment by DanielNoord** [comment]
> Right I forgot about that and just added it automatically to the next release. We need to create the 3.0 branch and deal with the 3.0 deprecations then.

We should probably create a separate issue for this discussion but I think we should at least make sure `astroid` can run `mypy`. After that we can be much more assured that the typing in `pylint` is correct and that we have collected all associated deprecations.
After that we can start thinking about any other new features we want to put in there.

**Comment by Felixoid** [comment]
`Needs review` label is redundant here now, right?

**Comment by Felixoid** [comment]
I am a bit confused, what is blocking the merge now?

**Comment by Pierre-Sassoulas** [comment]
Sorry for the delay @Felixoid. We need to create a 3.0 branch to merge and it's something that is going to create a lot of work if we need to cherry-pick everything on it. We have a lot to do with maintaining the 2.x branch as is, so it does not get done. We also have a ton of deprecation to handle before releasing 3.0.

**Comment by Felixoid** [comment]
It looks like now `.pylintrc` is not parsed unless `__init__.py` is found in the version 2.15.5 after https://github.com/PyCQA/pylint/pull/7423, and it breaks behavior for ClickHouse/ClickHouse. So, in the recent versions neither https://github.com/ClickHouse/ClickHouse/blob/a3e3be5/.pylintrc works nor pyproject.toml could be set up in the root directory.

Not sure if it was intended, but the current state is a quite strange from the configuration perspective.

**Comment by Pierre-Sassoulas** [comment]
Sorry for the delay again @Felixoid. The 3.0 release cycle is coming, are you still available to work on this or should we take over ?

**Comment by Felixoid** [comment]
Looks green as a coming spring!

**Comment by Felixoid** [comment]
Hello, are there any updates on it?..

File: doc/user_guide/usage/run.rst
Comment (DanielNoord) [inline doc/user_guide/usage/run.rst:109]:
```suggestion
   unless it's found, or a ``.git``/``.hg`` directory is found, or the file system root
```

**Review by DanielNoord — APPROVED** [review]
I think we can merge this, but defer to @Pierre-Sassoulas for doing so!

**Review by Pierre-Sassoulas — APPROVED** [review]
Sorry for taking a long time to review and merge, this is going to be in the first beta version ``3.0.0b1``

**Pull request merged**
